src/sync/app.cpp

```diff
@@ -81,11 +93,425 @@ static Optional<AppError> check_for_errors(const Response& response) {
+    if (!this->parent) {
+        return completion_block(Optional<UserAPIKey>(), AppError(make_custom_error_code(0), "UserAPIKeyProviderClient parent is null"));
+    }
+
+    std::string route = util::format("%1/auth/%2/%3", parent->m_base_route, m_provider_key, id.to_string());
+    
+    auto handler = [completion_block](const Response& response) {
+
+        if (auto error = check_for_errors(response)) {
+            return completion_block(Optional<UserAPIKey>(), error);
+        }
+                    
+        nlohmann::json json;
+        try {
+            json = nlohmann::json::parse(response.body);
+        } catch(const std::exception& e) {
+            return completion_block(Optional<UserAPIKey>(), AppError(make_error_code(JSONErrorCode::malformed_json), e.what()));
+        }
+        
+        try {
+            auto user_api_key = App::UserAPIKey {
+                    ObjectId(value_from_json<std::string>(json, "_id").c_str()),
+                    get_optional<std::string>(json, "key"),
+                    value_from_json<std::string>(json, "name"),
+                    value_from_json<bool>(json, "disabled")
+                };
+            return completion_block(user_api_key, {});
+        } catch(const std::exception& e) {
+            return completion_block(Optional<UserAPIKey>(), AppError(make_custom_error_code(0), e.what()));
+        }
+    };
+    
+    if (!SyncManager::shared().get_current_user()) {
+        return completion_block(Optional<UserAPIKey>(), AppError(make_custom_error_code(0), "SyncUser is null"));
+    }
+    
+    parent->m_config.transport_generator()->send_request_to_server({
+        HttpMethod::get,
+        route,
+        parent->m_request_timeout_ms,
+        get_request_headers(true),
+    }, handler);
+}
+
+void App::UserAPIKeyProviderClient::fetch_api_keys(std::function<void(std::vector<UserAPIKey>, Optional<AppError>)> completion_block) {
+    
+    if (!this->parent) {
+        return completion_block(std::vector<UserAPIKey>(), AppError(make_custom_error_code(0), "UserAPIKeyProviderClient parent is null"));
+    }
+
+    std::string route = util::format("%1/auth/%2", parent->m_base_route, m_provider_key);
+    
+    auto handler = [completion_block](const Response& response) {
+
+        if (auto error = check_for_errors(response)) {
+            return completion_block(std::vector<UserAPIKey>(), error);
+        }
+                    
+        nlohmann::json json;
+        try {
+            json = nlohmann::json::parse(response.body);
+        } catch(const std::exception& e) {
+            return completion_block(std::vector<UserAPIKey>(), AppError(make_error_code(JSONErrorCode::malformed_json), e.what()));
+        }
+
+        try {
+            auto api_key_array = std::vector<UserAPIKey>();
+            auto json_array = std::vector<nlohmann::json>(json);
+            for (nlohmann::json api_key_json : json_array) {
```

Comment:
This does a deep copy of each of the elements in the array while iterating over it; `for (nlohmann::json& api_key_json : json_array) {` to avoid that.

Reply:
Noted and addressed